sdk/Sdk.Generators/ExtensionStartupRunnerGenerator.cs

```diff
@@ -0,0 +1,212 @@
+        {
+            List<(string TypeName, string? Error)>? typeInfoList = null;
+
+            // Extension authors should decorate their assembly with "WorkerExtensionStartup" attribute
+            // if they want to participate in startup.
+            foreach (var assembly in context.Compilation.SourceModule.ReferencedAssemblySymbols)
+            {
+                var extensionStartupAttribute = assembly.GetAttributes()
+                                                        .FirstOrDefault(a => a.AttributeClass?.Name == attributeTypeName &&
+                                                                        //Call GetFullName only if class name matches.
+                                                                        a.AttributeClass.GetFullName() == attributeTypeFullName);
+                if (extensionStartupAttribute != null)
+                {
+                    // WorkerExtensionStartupAttribute has a constructor with one param, the type of startup implementation class.
+                    var firstConstructorParam = extensionStartupAttribute.ConstructorArguments[0];
+                    if (firstConstructorParam.Value is not ITypeSymbol typeSymbol)
+                    {
+                        continue;
+                    }
+
+                    var fullTypeName = typeSymbol.ToDisplayString();
+                    var errorMessage = GetErrorIfTypeIsNotValid(typeSymbol);
+
+                    typeInfoList ??= new List<(string TypeName, string? Error)>();
+                    typeInfoList.Add((TypeName: fullTypeName, Error: errorMessage));
+                }
+            }
+
+            return typeInfoList ?? Enumerable.Empty<(string TypeName, string? Error)>();
+        }
+
+        /// <summary>
+        /// Check the type to see it is valid and return an error message of the first failure.
+        /// </summary>
+        private string? GetErrorIfTypeIsNotValid(ITypeSymbol typeSymbol)
+        {
+            if (typeSymbol is INamedTypeSymbol namedTypeSymbol)
+            {
+                // Check public parameterless constructor exist for the type.
+                var constructorExist = namedTypeSymbol.InstanceConstructors
+                                                  .Any(c => c.Parameters.Length == 0 &&
+                                                            c.DeclaredAccessibility == Accessibility.Public);
+                if (!constructorExist)
+                {
+                    return $"{typeSymbol.ToDisplayString()} class must have a public parameterless constructor.";
+                }
+
+                // Check the extension startup class implements WorkerExtensionStartup abstract class.
+                if (!(namedTypeSymbol.BaseType!.GetFullName().Equals(startupBaseClassName)))
+                {
+                    return $"{typeSymbol.ToDisplayString()} must be a type which implements {startupBaseClassName}.";
```

Comment:
nit: Change to `$"{typeSymbol.ToDisplayString()} must derive from {startupBaseClassName}."`

Reply:
Changed the language as suggested.